unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,574 @@
+            )
+        except NodeIDAbsentError:
+            pass
+        except DuplicatedNodeIdError:
+            pass
+
+    def get_entry(self, object_id: int):
+        try:
+            entry = self.file_entries.get_node(object_id)
+            if entry:
+                return entry.data
+        except NodeIDAbsentError:
+            pass
+        return None
+
+    def resolve_path(self, entry: YAFFSEntry) -> Path:
+        resolved_path = Path(entry.name)
+        parent = self.file_entries[entry.parent_obj_id].data
+        if parent is not None:
+            return self.resolve_path(parent).joinpath(resolved_path)
+        return resolved_path
+
+    def get_file_bytes(self, entry: YAFFSEntry) -> Iterable[bytes]:
+        for chunk in entry.get_chunks():
+            self.file.seek(
+                entry.start_offset
+                + (
+                    (chunk.chunk_id - 1)
+                    * (self.config.page_size + self.config.spare_size)
+                ),
+                io.SEEK_SET,
+            )
+            yield self.file.read(chunk.byte_count)
+
+    def extract(self, outdir: Path):
+        self.file_entries.show()
+        for entry in [
+            self.file_entries.get_node(node).data
+            for node in self.file_entries.expand_tree(mode=Tree.DEPTH)
+        ]:
+            if entry is None:
+                continue
+            self.extract_entry(entry, outdir)
+
+    def extract_entry(self, entry: YAFFS2Entry, outdir: Path):  # noqa: C901
+        entry_path = self.resolve_path(entry)
+
+        if not is_safe_path(outdir, entry_path):
+            logger.warning(
+                "Potential path traversal attempt", outdir=outdir, path=entry_path
+            )
+            return
+
+        out_path = outdir.joinpath(entry_path)
+
+        if entry.object_type == YaffsObjectType.DIRECTORY:
+            logger.debug("creating directory", dir_path=out_path, _verbosity=3)
+            out_path.mkdir(exist_ok=True)
+        elif entry.object_type == YaffsObjectType.FILE:
+            logger.debug("creating file", file_path=out_path, _verbosity=3)
+            with out_path.open("wb") as f:
+                for chunk in self.get_file_bytes(entry):
+                    f.write(chunk)
+        elif entry.object_type == YaffsObjectType.SPECIAL:
+            if os.geteuid() == 0:
+                logger.debug(
+                    "creating special file", special_path=out_path, _verbosity=3
+                )
+                os.mknod(out_path.as_posix(), entry.yst_mode, entry.yst_rdev)
+            else:
+                logger.warning(
+                    "creating special files requires elevated privileges, skipping.",
+                    path=out_path,
+                    yst_mode=entry.yst_mode,
+                    yst_rdev=entry.yst_rdev,
+                )
+        elif entry.object_type == YaffsObjectType.SYMLINK:
+            if not is_safe_path(outdir, out_path / Path(entry.alias)):
```

Comment:
wondering now, that this would not be reported as MaliciousSymlinkRemoved report, while other extractors would

Also the above path-traversal is just logger, but not reported..

Reply:
This is something that must absolutely be handled within the extractor. If we don't we're vulnerable to this attack:

1. create a symlink entry linking outside (`/home/extract/a -> /etc/passd`)
2. create a file entry with the same name (`/home/extract/a`)

When the file entry gets extracted, it will write out to the symlink target.

Reply:
This is probably something we should look for in our other handlers by the way.

Reply:
TODO @qkaiser create a ticket for this

Reply:
Done. https://github.com/onekey-sec/unblob/issues/550